airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/InputConsumerTask.kt

```diff
@@ -100,6 +101,43 @@ class InputConsumerTask(
                     }
                     unopenedStreams
                 }
+            } catch (e: Exception) {
+                log.warn(e) {
+                    "Encountered exception in input loop. Capturing exception and proceeding with sync shutdown."
+                }
+                baseException = e
+                // intentionally don't throw here.
+                // we'll throw an appropriate exception in the `finally` block.
+            } finally {
+                // pipelineEventBookkeepingRouter.use {...} will swallow baseException.
+                // so instead, we'll use a plain try-catch-finally,
+                // and actually prioritize the original exception over anything thrown
+                // in this `try` block.
+                try {
+                    // this is likely to throw a "no stream status message" error,
+                    // if there's an error during input handling.
+                    // That's quite misleading (b/c the actual error is that we crashed before we
+                    // could handle the stream status message),
+                    // so we would prefer to rethrow the original exception.
+                    // but we can't just skip this close() call if baseException != null,
+                    // because that would cause the sync to hang.
+                    pipelineEventBookkeepingRouter.close()
+                } catch (e: Exception) {
+                    if (baseException == null) {
+                        throw e
+                    } else {
+                        log.warn(e) {
+                            "Encountered exception when closing PipelineEventBookkeepingRouter, but we're already handling another exception ($baseException). Swallowing this exception."
+                        }
+                        throw baseException
+                    }
+                }
+                // if the `pipelineEventBookkeepingRouter.close()` ran successfully,
+                // we should rethrow the original exception.
+                if (baseException != null) {
+                    log.warn { "Throwing captured exception after triggering sync shutdown." }
+                    throw baseException
+                }
```

Comment:
This logic looks fine to me, but I'd strongly consider moving it to a separate method and unit testing it to validate it works the way we expect

Reply:
good call. I'll also add a case to the Mock test, in case some other CDK layer _also_ swallows the exception (i.e. enforce the actual behavior we care about, which is that we emit a trace message with correct information)

Reply:
added unit+integration tests, PTAL